### PR TITLE
runtime: hashmap cleanups

### PR DESCRIPTION
**Pull request by dgryski**

### CODE REVIEW

File: src/runtime/hashmap.go [hunk 1]
```diff
@@ -154,34 +153,61 @@ func hashmapLenUnsafePointer(m unsafe.Pointer) int {
...
+	return slotKey
+}
+
+//go:inline
+func hashmapSlotValue(m *hashmap, bucket *hashmapBucket, slot uint8) unsafe.Pointer {
+	slotValueOffset := unsafe.Sizeof(hashmapBucket{}) + uintptr(m.keySize)*8 + uintptr(m.valueSize)*uintptr(slot)
+	slotValue := unsafe.Pointer(uintptr(unsafe.Pointer(bucket)) + slotValueOffset)
+	return slotValue
+}
+
```
Comment (soypat) [inline src/runtime/hashmap.go:188]: I appreciate you adding these helper functions, makes the code a lot more readable

File: src/runtime/hashmap.go [hunk 2]
```diff
@@ -154,34 +153,61 @@ func hashmapLenUnsafePointer(m unsafe.Pointer) int {
 	return hashmapLen((*hashmap)(m))
 }
 
+//go:inline
```
Comment (soypat) [inline src/runtime/hashmap.go:156]: Is there a straightforward way to check change in binary size by looking at the CI? I checked https://github.com/tinygo-org/tinygo/actions/runs/4527779105/jobs/7974008883?pr=3605 but could not find a diff inside the log.

**Review by soypat — COMMENTED** [review]
LGTM- first time I'm looking at hashmap code. Looks fun

**Review by deadprogram — APPROVED** [review]
Thanks for the nice commit atomicity. Looks good to me, although I would certainly defer to @aykevl or others on the rigor of the implementation.

File: src/runtime/hashmap.go [hunk 3]
```diff
@@ -154,34 +153,61 @@ func hashmapLenUnsafePointer(m unsafe.Pointer) int {
...
+func hashmapBucketAddrForHash(m *hashmap, hash uint32) *hashmapBucket {
+	numBuckets := uintptr(1) << m.bucketBits
+	bucketNumber := (uintptr(hash) & (numBuckets - 1))
+	return hashmapBucketAddr(m, m.buckets, bucketNumber)
+}
+
+//go:inline
+func hashmapSlotKey(m *hashmap, bucket *hashmapBucket, slot uint8) unsafe.Pointer {
+	slotKeyOffset := unsafe.Sizeof(hashmapBucket{}) + uintptr(m.keySize)*uintptr(slot)
+	slotKey := unsafe.Pointer(uintptr(unsafe.Pointer(bucket)) + slotKeyOffset)
```
Comment (kenbell) [inline src/runtime/hashmap.go]: Should use `unsafe.Add` ?

File: src/runtime/hashmap.go [hunk 4]
```diff
@@ -154,34 +153,61 @@ func hashmapLenUnsafePointer(m unsafe.Pointer) int {
...
+	return slotKey
+}
+
+//go:inline
+func hashmapSlotValue(m *hashmap, bucket *hashmapBucket, slot uint8) unsafe.Pointer {
+	slotValueOffset := unsafe.Sizeof(hashmapBucket{}) + uintptr(m.keySize)*8 + uintptr(m.valueSize)*uintptr(slot)
+	slotValue := unsafe.Pointer(uintptr(unsafe.Pointer(bucket)) + slotValueOffset)
```
Comment (kenbell) [inline src/runtime/hashmap.go]: Should use `unsafe.Add` here also?

**Review by kenbell — APPROVED** [review]
Couple of nits

**Comment by dgryski** [comment]
PTAL

**Comment by deadprogram** [comment]
Thanks for all the cleanups @dgryski and thanks @kenbell and @soypat for helping review. Now merging.

**Pull request merged**